Chatkit/View Models/JoinedRoomsViewModel.swift

```diff
@@ -8,195 +8,95 @@ import Foundation
 ///
 /// ## What is provided
 ///
-/// The ViewModel exposes an array, `rooms: [Room]` which presents the rooms that the current user is a member
-/// of in descending order of the time of their last message, or their creation time if they contain no messages.
+/// The ViewModel exposes  a state with an array of `rooms` which presents the rooms that the current
+/// user is a member of in descending order of the time of their last message, or their creation time if they
+/// contain no messages.
 ///
-/// Each item in the `rooms` array can be used to populate a cell in a `UITableView` or `UICollectionView`.
+/// Each item in the `rooms` array can be used to populate a cell in a `UITableView`
+/// or `UICollectionView`.
 ///
 /// ## Receiving live updates
 ///
-/// In order to be notified when the contents of the `rooms` changes, implement the `JoinedRoomsViewModelDelegate` protocol and assign the `JoinedRoomsViewModel.delegate` property.
-///
-/// Note that when the view model is first returned to you, it will already be populated, and the delegate will only be invoked when the contents change.
+/// In order to be notified when the contents of the `rooms` or the `state` of the connection changes,
+/// implement the `JoinedRoomsViewModelDelegate` protocol and assign
+/// the `JoinedRoomsViewModel.delegate` property.
 ///
 /// ## Understanding the `state` of the ViewModel
 ///
 /// The `state` property describes the state of the live update connection, either
+///   - `.initializing`: awaiting the initial set of data, or
 ///   - `.connected`: updates are flowing live, or
-///   - `.degraded`: updates may be delayed due to network problems.
-///
+///   - `.degraded`: updates may be delayed due to network problems, or
+///   - `.closed`: the connection is closed, no further updates available.
 public class JoinedRoomsViewModel {
     
     // MARK: - Properties
     
-    private let provider: JoinedRoomsProvider
+    private let repository: JoinedRoomsRepositoryProtocol
     
-    /// The array of all rooms joined by the user.
-    public private(set) var rooms: [Room]
-    
-    /// The current state of the provider used by the view model as the data source.
-    public var state: RealTimeProviderState {
-        return self.provider.state
+    /// The current state of the repository.
+    public private(set) var state: State {
+        didSet {
+            if state != oldValue {
+                self.delegate?.joinedRoomsViewModel(self, didUpdateState: state)
```

Comment:
The delegate of a `JoinedRoomsViewModel` will always be a client entity and therefore we should always invoke this on the main queue in case the state change came from a background queue...
```suggestion
                DispatchQueue.main.async {
                    self.delegate?.joinedRoomsViewModel(self, didUpdateState: self.state)
                }

```